File: Tut_07/Java/Stack_LinkedList.java
package Tut_07;


public class Stack_LinkedList <T> {
	Node <T> head = null;
	// print the item
	public void printList() {
		Node temp = this.head;
		while (temp != null) {
			System.out.print(temp.data + " ");
			temp = temp.next;
		}
	}

	// isEmpty(): Check if the stack is empty
	public boolean isEmpty() {
		if (this.head == null)
			return true;
		return false;
	}

	// push(itemType newItem): Add a new item to the stack
	public void push(char c) {

		// Create a new node which data = c
		Node newNode = new Node(c);

		// chang the head to it
		newNode.next = this.head;
		this.head = newNode;
	}

	// pop(): Take out an item from the stack
	public void pop() {
		while (this.isEmpty()) {
			System.out.println("Nothing left to pop");
			break;
		}
		
		// Take the head out
		this.head = this.head.next;
	}
	
	//  peek(): Retrive an item from the stack (without deleting it)
	public char peek() {
		if ( this.isEmpty())
			System.out.println("Error. Nothing left to pop");
		
		
		return (char) this.head.data;
	}
	

}

File: Tut_07/Python/stack_linkedlist.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Stack:
    def __init__(self):
        # Node head
        self.head = None


    def printList(self):
        if self.head is None:
            return 'There is nothing left'
        temp = self.head
        while temp is not None:
            print(temp.data)
            temp = temp.next


    def isEmpty(self):
        return self.head == None


    def push(self, new_data):
        # create a new node
        new_Node = Node(new_data)

        new_Node.next = self.head
        self.head = new_Node

    def pop(self):
        self.head = self.head.next


    def peek(self):
        temp = self.head
        while temp.next is not None:
            temp = temp.next
        print(temp.data)

    def delete(self):
        self.head = None





if __name__ == '__main__':
    stack = Stack()
    stack.push(1)
    stack.push(2)
    stack.push(3)
    stack.pop()
    print('Last time')
    stack.printList()

